### PR TITLE
Port Toplevel to the Stm API

### PR DESCRIPTION
**Pull request by ejgallego**
As discussed with Enrico, we extend the Stm API to decouple parsing from document building, and we port the toplevel to the new API removing quite a bit of redundancy. A few minor cleanups happen in the way.

This greatly helps clients willing to do whole-document parsing, among other things. pidecoq was ported without problems that I could spot.

The PR is ready to merge, and it supersedes #203 and #204 .

### CODE REVIEW

File: ide/ide_slave.ml [hunk 1]
```diff
@@ -78,56 +79,42 @@ let is_known_option cmd = match cmd with
...
 (** Interpretation (cf. [Ide_intf.interp]) *)
 
 let add ((s,eid),(sid,verbose)) =
-  let newid, rc = Stm.add ~ontop:sid verbose ~check:coqide_cmd_checks eid s in
-  let rc = match rc with `NewTip -> CSig.Inl () | `Unfocus id -> CSig.Inr id in
+  let pa      = Pcoq.Gram.parsable (Stream.of_string s) in
+  let loc_ast = Stm.parse_sentence sid pa               in
+  coqide_cmd_checks loc_ast;
+  let newid, rc = Stm.add ~ontop:sid verbose loc_ast in
+  let rc        = match rc with `NewTip -> CSig.Inl () | `Unfocus id -> CSig.Inr id in
```
Comment (Zimmi48) [inline ide/ide_slave.ml]: I don't like too much this internal indentation. I don't think it brings much clarity; it contrasts with the rest of the code-base; it makes some lines longer; and it makes git diff less clean (for instance let's say that I want to change `newid` to `new_id`, I would then have to add a new space on the line below to keep indentation consistent).

Of course this is just MVHO; I'm not the one to decide on the coding style; and I might have missed some earlier changes of yours which could have brought that style in the code-base already, in which case my second argument does not hold anymore.

File: ide/xmlprotocol.ml [hunk 2]
```diff
@@ -871,13 +869,9 @@ let of_feedback msg =
...
 let to_feedback xml = match xml with
-  | Element ("feedback", ["object","edit";"route",route], [id;content]) -> {
-      id = Edit(to_edit_id id);
-      route = int_of_string route;
-      contents = to_feedback_content content }
   | Element ("feedback", ["object","state";"route",route], [id;content]) -> { 
-      id = State(to_stateid id);
-      route = int_of_string route;
+      id       = to_stateid id;
+      route    = int_of_string route;
```
Comment (Zimmi48) [inline ide/xmlprotocol.ml]: I have the same comment here. And this illustrates one of my argument. Would you not worry about internal indentation, you would have changed one line here instead of two.

File: ide/ide_slave.ml (hunk 1 above)
Comment (gares) [inline ide/ide_slave.ml]: I agree, if one wants to keep this style he would have to change lines (adding spaces) not affected by his patch.

File: stm/stm.mli [hunk 3]
```diff
@@ -182,14 +189,9 @@ val register_proof_block_delimiter :
...
 val state_computed_hook : (Stateid.t -> in_cache:bool -> unit) Hook.t
-val parse_error_hook :
-  (Feedback.edit_or_state_id -> Loc.t -> Pp.std_ppcmds -> unit) Hook.t
-val unreachable_state_hook : (Stateid.t -> Exninfo.iexn -> unit) Hook.t
+
 (* ready means that master has it at hand *)
 val state_ready_hook : (Stateid.t -> unit) Hook.t
-(* called with true before and with false after a tactic explicitly
- * in the document is run *)
-val tactic_being_run_hook : (bool -> unit) Hook.t
```
Comment (gares) [inline stm/stm.mli]: This was needed by the XML-Export plugin.  Not sure we ever received such code from Claudio.  Such code was in turn a which of Airbus IIRC... To be evaluated before merge (since orthogonal to this patch AFAICT)

File: ide/ide_slave.ml (hunk 1 above)
Comment (ejgallego) [inline ide/ide_slave.ml]: Internal indentation makes the code demonstrably easier to read and reason about, but indeed diff pollution is a big price to pay, so I try to introduce it only in new code, using a non-strict style. Fixed on both counts.

File: toplevel/coqloop.ml [hunk 4]
```diff
@@ -337,11 +336,15 @@ let loop_flush_all () =
 
 let rec loop () =
   Sys.catch_break true;
-  if !Flags.print_emacs then Vernacentries.qed_display_script := false;
-  Flags.coqtop_ui := true;
   try
     reset_input_buffer stdin top_buffer;
-    while true do do_vernac(); loop_flush_all () done
+    (* Tail-rec is essential *)
+    let rec loop sid =
```
Comment (gares) [inline toplevel/coqloop.ml]: This inner loop should not have the name of the outer loop.

My personal taste leans toward
```
let cur := ref (Stm.get_current_state()) in
while true do cur := do_vernac !cur; flush(); done
```
I mean, if it has to be a loop, let it be a real loop ;-) ... I know, I'm how a Haskell guy...

File: stm/stm.mli (hunk 3 above)
Comment (ejgallego) [inline stm/stm.mli]: This PR is really about cruft removal so this is why that was included there. IMHO we have accumulated too much cruft and duplication in the codebase, 8.7 could be a good opportunity to clear things ups a bit.

That being said, just let me know what you think and I'll drop that patch.

File: toplevel/coqloop.ml (hunk 4 above)
Comment (ejgallego) [inline toplevel/coqloop.ml]: I will change the name, but sorry, the loop there is a real loop and IMVHO we should avoid using references gratuitously.

File: toplevel/coqinit.ml [hunk 5]
```diff
@@ -43,8 +43,8 @@ let load_rcfile() =
 	    Envars.home ~warn / "."^rcdefaultname^"."^Coq_config.version;
 	    Envars.home ~warn / "."^rcdefaultname
 	  ] in
-          Vernac.load_vernac false inferedrc
-	with Not_found -> ()
+          Vernac.load_vernac Stateid.initial inferedrc
```
Comment (gares) [inline toplevel/coqinit.ml]: why `initial` and not `sid`?

Comment (ejgallego) [inline toplevel/coqinit.ml]: Good catch, thanks!

File: vernac/vernacprop.ml [hunk 6]
```diff
@@ -0,0 +1,53 @@
+(************************************************************************)
```
Comment (gares) [inline vernac/vernacprop.ml]: shouldn't this code end up ion vernac_classifier?  Yes, it is not exactly the same output (bool v.s. something more fine grained) but semantically this is a classafication...

File: stm/stm.ml [hunk 7]
```diff
@@ -370,6 +321,12 @@ end = struct (* {{{ *)
   open Printf
 
   let print_dag vcs () =
+
+    (* Due to threading, this will be called from the toplevel with
+       we_are_parsing set to true, as indeed, the toplevel is waiting
+       for input . What a race! XD *)
+    Flags.we_are_parsing := false;
```
Comment (gares) [inline stm/stm.ml]: I'm not following.  This boolean is there to detect "commands" that access the system state (the global environment) while parsing.  I.e. commands that behave differently if I parse;parse;exec;exec or parse;exec;parse;exec.

Why are you flipping this bit here?

File: vernac/vernacprop.ml (hunk 6 above)
Comment (ejgallego) [inline vernac/vernacprop.ml]: IMHO it belong to `vernac`  more than `stm`.

File: stm/stm.ml (hunk 7 above)
Comment (ejgallego) [inline stm/stm.ml]: The race is like this:
- The stm schedules the plotting in an ocaml thread, it gets queued.
- The toplevel calls `Stm.parse_sentence` with a parseable connected to `stdin`.
- `Stm.parse_sentence` sets the `we_are_parsing` flag to true.
- Reading from the console is a blocks, with the flag on, now the pending DAG drawing thread will be scheduled.
- Boom!

A good lesson on the dangers of imperative flags :D

**Comment by gares** [comment]
I still have to review the patch on stm.ml (and I'm going home now ;-)
The rest is good stuff!

File: stm/stm.ml (hunk 7 above)
Comment (gares) [inline stm/stm.ml]: Oh, you need to access the global environment to print?  Shit...

Comment (gares) [inline stm/stm.ml]: Anyway, Dag.print is for debugging only... but flipping the boolean is wrong.  What you want to do is to catch the exception and stop printing in that case.  Let's call it "best effort debug print"

**Comment by ejgallego** [comment]
thanks a lot for the review @gares

File: ide/ide_slave.ml (hunk 1 above)
Comment (Zimmi48) [inline ide/ide_slave.ml]: I'll digress a bit, but I've been influenced lately by the radical view of Elm w.r.t indenting style (cf. http://elm-lang.org/docs/style-guide). They even have an automatic formatter on save (https://github.com/avh4/elm-format) to avoid thinking about it and keep a consistent style within a project and even across projects. I have the feeling that there are some good ideas to borrow.

File: stm/stm.ml (hunk 7 above)
Comment (ejgallego) [inline stm/stm.ml]: Then you won't be able to print anything. You are right int that the boolean should be restored after printing.

Or even better, the boolean should be totally obliterated, but that belongs to a different discussion.

File: ide/ide_slave.ml (hunk 1 above)
Comment (ejgallego) [inline ide/ide_slave.ml]: YMMV

Comment (ejgallego) [inline ide/ide_slave.ml]: [Most of my diff tools will ignore whitespace changes by the way]

File: stm/stm.mli (hunk 3 above)
Comment (gares) [inline stm/stm.mli]: @maximedenes was in the loop for XML-Export

File: vernac/vernacprop.ml (hunk 6 above)
Comment (gares) [inline vernac/vernacprop.ml]: vernac_classifier is in the same position, it belongs to vernac given how it is named, bet needs to be accessible by Stm.  Maybe the vernac/ directory deserves a better name.  It contain the main interpreter.... I'm all ears

Comment (ejgallego) [inline vernac/vernacprop.ml]: Yeah not sure, maybe if vernac_classifier needs access to the document then it belongs in stm.

Comment (gares) [inline vernac/vernacprop.ml]: My point is that, not matter where vernac_classifies is or vernacprop is, they do very similar things: group vernac commands into classes.

Comment (ejgallego) [inline vernac/vernacprop.ml]: I see what you mean, however vernac_classifier currently needs access to the VCS whereas the simpler classifier does not.

**Comment by herbelin** [comment]
Is it transparent for the coqtop user, or does this imply changes of behavior?

**Comment by ejgallego** [comment]
Hi Hugo, the idea is that is 100% transparent, unless explicitly desired otherwise.

However, there could exist some small discrepancies: some of the old logic in the `Stm` for the case `tty:true` was a bit convoluted and hard to follow. I am quite confident that if testing reveals such a problem it would be easy to fix.

**Comment by herbelin** [comment]
OK, thanks for the answer.

**Comment by ejgallego** [comment]
Ok this has been rebased, and each individual commit tested for building. Lets see what Travis thinks, again this should be ready to go modulo objections. So far we have:
- @gares points out about the location of the Vernacprop file. I think given the current structure it makes sense to have them separate, but no strong opinion on that:
- @maximedenes @gares point out about whether the `tactic_being_run` hook should be preserved. Will do as instructed.

coqPIDEtop seems to work fine with this patch, tree at https://bitbucket.org/ejgallego/pidetop/branch/v8.7%2Bedit_id

File: dev/doc/changes.txt [hunk 8]
```diff
@@ -128,6 +128,36 @@ document type". This allows for a more uniform handling of printing
...
+** Stm API **
+
+- We have streamlined the `Stm` API, now `add` and `query` take a
+  `coq_parsable` instead a `string` so clients can have more control
+  over their input stream. As a consequence, their types have been
+  modified.
+
+- The main parsing entry point has also been moved to the
+  `Stm`. Parsing is considered a synchronous operation so it will
+  either succeed of emit an exception.
```
Comment (psteckler) [inline dev/doc/changes.txt]: typo: *or* emit an exception.

File: dev/doc/changes.txt [hunk 9]
```diff
@@ -128,6 +128,36 @@ document type". This allows for a more uniform handling of printing
...
+- The components related to vernacular interpretation have been moved
+  to their own folder `vernac/` whereas toplevel now contains the
+  proper toplevel shell and compiler.
+
+- Coq's toplevel has been ported to directly use the common `Stm`
+  API. The signature of a few functions has changed as a result.
+
+** XML Protocol **
+
+- The legacy `interp` call has been turned into a noop.
```
Comment (psteckler) [inline dev/doc/changes.txt]: Should this be capitalized: *Interp*, since that's what goes in the XML, AFAIK.

File: dev/doc/changes.txt (hunk 8 above)
Comment (ejgallego) [inline dev/doc/changes.txt]: Cool will fix.

File: dev/doc/changes.txt (hunk 9 above)
Comment (ejgallego) [inline dev/doc/changes.txt]: Cool will fix.

**Comment by ejgallego** [comment]
Ok, fd46bb5 pushed, thanks @psteckler , let me know what you think of the messages so I can amend the patch if needed.

**Comment by psteckler** [comment]
Looks good, thanks!

**Comment by maximedenes** [comment]
@gares Could you review the last bit in `stm.ml`? Thanks!

File: stm/stm.ml [hunk 10]
```diff
@@ -369,6 +320,13 @@ end = struct (* {{{ *)
   open Printf
 
   let print_dag vcs () =
+
+    (* Due to threading, this will be called from the toplevel with
+       we_are_parsing set to true, as indeed, the toplevel is waiting
+       for input . What a race! XD *)
+    let old_we_are_parsing = !Flags.we_are_parsing in
```
Comment (maximedenes) [inline stm/stm.ml]: remove that. it does not fix the race

File: stm/stm.ml [hunk 11]
```diff
@@ -882,7 +844,6 @@ end = struct (* {{{ *)
         | Some _, None -> (e, info)
         | Some (_,at), Some id -> (e, Stateid.add info ~valid:id at) in
       if cache = `Yes || cache = `Shallow then freeze_invalid id ie;
-      Hooks.(call unreachable_state id ie);
```
Comment (maximedenes) [inline stm/stm.ml]: This is used by PIDE, your patch to pidetop does remove the hook call but it is needed.
It is like the `cancel` messages of serapi, PIDE needs to be notifies of each and every state (exec id) that is processed (or cancelled for processing).

File: stm/stm.ml [hunk 12]
```diff
@@ -2390,20 +2346,20 @@ let observe id =
...
   observe (VCS.get_branch_pos head);
-  if print_goals then msg_notice (pr_open_cur_subgoals ());
   VCS.print ();
+  (* EJGA: Setting here the proof state looks really wrong, and it
+     hides true bugs cf bug #5363. Also, what happens with observe? *)
   (* Some commands may by side effect change the proof mode *)
   match VCS.get_branch head with
   | { VCS.kind = `Edit (mode,_,_,_,_) } -> Proof_global.activate_proof_mode mode
-  | { VCS.kind = `Proof (mode, _) } -> Proof_global.activate_proof_mode mode
+  | { VCS.kind = `Proof (mode, _) }     -> Proof_global.activate_proof_mode mode
```
Comment (maximedenes) [inline stm/stm.ml]: nope

File: stm/stm.ml [hunk 13]
```diff
@@ -2713,29 +2650,91 @@ let get_ast id =
...
+  else match (VCS.visit sid).step with
+  | `Cmd { ctac = true; cast = { indentation; strlen; loc } } ->
+         Some (indentation, strlen, loc)
+  | _ -> None
+
+(* the indentation logic works like this: if the beginning of the
+   command is different than the start we are in a new line, thus
+   indentation follows from the starting position. Otherwise, we use
+   the previous one plus the offset. *)
+(* XXX: Note, this can be surely improved *)
```
Comment (maximedenes) [inline stm/stm.ml]: how

File: stm/stm.ml [hunk 14]
```diff
@@ -2751,17 +2750,16 @@ let query ~at ?(report_with=(Stateid.dummy,default_route)) s =
   Future.purify (fun s ->
     if Stateid.equal at Stateid.dummy then finish ()
     else Reach.known_state ~cache:`Yes at;
-    let newtip, route = report_with in
-    let indentation, strlen, loc, ast = vernac_parse ~newtip ~route 0 s in
+    let loc, ast            = parse_sentence at s in
```
Comment (maximedenes) [inline stm/stm.ml]: nope

File: toplevel/coqtop.ml [hunk 15]
```diff
@@ -634,15 +630,16 @@ let init_toplevel arglist =
...
       then Declaremods.start_library !toplevel_name;
       init_library_roots ();
       load_vernac_obj ();
       require ();
       Stm.init ();
-      load_rcfile();
-      load_vernacular ();
+      let sid  = load_rcfile (Stm.get_current_state ()) in
+      (* XXX: We ignore this for now, but should be threaded to the toplevels *)
+      let _sid = load_vernacular sid                       in
```
Comment (maximedenes) [inline toplevel/coqtop.ml]: nope

File: dev/doc/changes.txt [hunk 16]
```diff
@@ -139,6 +139,36 @@ document type". This allows for a more uniform handling of printing
...
+- The components related to vernacular interpretation have been moved
+  to their own folder `vernac/` whereas toplevel now contains the
+  proper toplevel shell and compiler.
+
+- Coq's toplevel has been ported to directly use the common `Stm`
+  API. The signature of a few functions has changed as a result.
+
+** XML Protocol **
+
+- The legacy `Interp` call has been turned into a noop.
```
Comment (maximedenes) [inline dev/doc/changes.txt]: I think it should give an error, like `Ùnsupported Protocol Message`

**Review by maximedenes — CHANGES_REQUESTED** [review]

**Comment by ejgallego** [comment]
Comments fixed, thanks @gares !
